enhancements/oc/mirroring-manifests.md

```diff
@@ -0,0 +1,172 @@
+approvers:
+  - "@LalatenduMohanty"
+  - "@smarterclayton"
+creation-date: 2020-01-24
+last-updated: 2020-01-24
+status: provisional|implementable
+---
+
+# Mirroring Manifests
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+For disconnected clusters, `oc adm release mirror ...` already helps users copy the release image and its dependencies into their local mirror.
+This enhancment extends that command to also create and apply manifests to their local cluster, to bring in information that cannot be represented by container images.
+For example, `oc` might produce ConfigMap manifests containing the release image signature which the cluster-version operator could use to [verify the mirrored release][cvo-config-map-signatures].
+
+## Motivation
+
+There is a lot of information in the wider internet that may be useful for restricted-network clusters.
+Some of that information can be represented by container images, such as the release image and the operator and other images referenced by the release image.
+Some of that information cannot be represented by container images, including container signatures.
+By extending `oc`'s release-mirror support to create and apply Kubernetes manifests, we provide a channel that can be used to gather arbitrary information in one place that the user can audit, transport into their restricted network, and apply to their local cluster.
+As the need for additional information arises, additional manifests may be created, possibly using new [custom resources][custom-resources] known to the local cluster, without users needing to adjust their mirroring workflows.
+
+### Goals
+
+Users should be able to run a single `oc adm release mirror ...` command to gather all the external information needed to deploy that release inside their cluster, which may have restricted network access.
+
+### Non-Goals
+
+This proposal does not specify the types of manifests that are created.
+Users who intend to audit the gathered manifests should extract the gathered types from the manifests themselves.
+
+## Proposal
+
+### Applying directly to the target cluster
+
+As of 4.2 and 4.3:
+
+```console
+$ oc adm release mirror \
+>   --from=quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64 \
+>   --to=registry.example.com/your/repository \
+>   --to-release-image=registry.example.com/your/repository:your-tag
+```
+
+will mirror the 4.3.0 release image and other images referenced by the release image into the mirror repository at registry.example.com.
+
+This enhancement would also apply associated manifests to the cluster that `oc` connects to, with the manifest application beginning after the image mirror had completed.
```

Comment:
What are the odds that someone actually has access to the machine in a restricted network?  This wouldn't help most disconnected customers as much.  So it's possible that this just isn't the right first step (but thinking through the next step might be fine).

Might be good to break this up into three parts:

1. mirroring and outputting a set of manifests that can be applied to the server
2. mirroring and outputting and applying to a target server directly
3. a description of the human flow for mirroring to a central registry in a fully air gapped environment

instead of the two sections below.

Reply:
> What are the odds that someone actually has access to the machine in a restricted network?

Slim, but you may decide to mirror to clusters without having a restricted network, e.g. to reduce exposure to RH outages or whatever.

> mirroring and outputting a set of manifests that can be applied to the server

Is this "pushing images directly to a different registry, but dropping manifests to disk, because I need to sneakernet those in"?  I guess we can do that, but it doesn't seem all that different from mirroring to disk locally, turning around and pushing the images immediately with `oc image mirror ...`, and then sneakernetting the manifests in to `oc apply` them.

Reply:
> ...mirroring to disk locally, turning around and pushing the images immediately with oc image mirror ..., and then sneakernetting the manifests in to oc apply them.

This is confusing. If we have to sneakernet the manifests then one has to sneakernet everything - including the images.

Reply:
> If we have to sneakernet the manifests then one has to sneakernet everything - including the images.

No, [the 4.2 docs][1] (which predate `--to-dir`) say:

> In this procedure, you place the mirror registry on a bastion host that has access to both your network and the internet. If you do not have access to a bastion host, use the method that best fits your restrictions to bring the contents of the mirror registry into your restricted network.

So there are some folks with a local registry accessible from inside and outside, but a cluster that cannot reach outside.  Or that was just the best story we had for 4.2, even if there were no users in that situation?

[1]: https://docs.openshift.com/container-platform/4.2/installing/install_config/installing-restricted-networks-preparations.html

Reply:
When I see the term sneakernet, and therefore the necessity to do so, we must be dealing with an "air-gapped", i.e. physically isolated, network and it can't be connected to anything that is connected to the internet. If it isn't air-gapped, and is connected to the outside in some way, shape or form, there would be no need to sneakernet so why would we.

Reply:
> When I see the term sneakernet...

Our official openshift-docs use "restricted network" instead of "disconnected", "sneakernet", etc. specifically because there are a whole range of possible setups in this space.  You might want to use `oc adm release mirror ...` to mirror into your local space in any of those, and maybe even if you have unrestricted access to the wider internet but have, say, policies limiting a running cluster's reliance on assets you do not directly control.

Reply:
> > mirroring and outputting a set of manifests that can be applied to the server
>
> Is this "pushing images directly to a different registry, but dropping manifests to disk, because I need to sneakernet those in"? I guess we can do that...

Talking to @mazzystr about this enhancement today, I'm going to reverse my earlier "hrm..." to be "yeah, we want this without a separate command", because it is how install is going to work when the mirror registry is available for immediate container image pushing.  Updated the enhancement to add a new `--manifests-to-dir` with 8d55831 -> 9517e1f.